feat: social tips contract (#115)
desmos-labs/desmos-contracts: docs/architecture/adr-005-social-tips-contract.md

### Messages

#### Instantiate
```rust
pub struct InstantiateMsg {
    pub admin: Option<String>,
    pub max_pending_tips: u16,
    pub max_sent_pending_tips: u16,
}
```

> * The `admin` identifies the user that controls the contract, if is `None` the contract will pick the address of who sent the transaction;

Reviewer: ```suggestion
* The `admin` identifies the user that controls the contract, if it is `None` the contract will pick the address of who sent the transaction;
```

* the funds necessary to cover the fees plus the tip amount specified inside the `funds` field
* the sender identified by the `sender` field

> **NOTE** If the user have already sent a tip to the same centralized application reference the contracts collects

Reviewer: ```suggestion
**NOTE** If the user has already sent a tip to the same centralized application reference the contracts collects
```

> only the last one sending back to the user the previous tip amount.

Reviewer: ```suggestion
only the last one then sending back to the user the previous tip amount.
```

#### ClaimTips
With the `ClaimTips` message the user can claim their pending tips that has been sent to him before proving that
owns the identities to which the tips refer.

##### UpdateAdmin

> With the `UpdateAdmin` message the current admin can update the contract's admin.

Reviewer: ```suggestion
With the `UpdateAdmin` message, the current admin can update the contract's admin.
```

### UpdateMaxPendingTips

> With the `UpdateMaxPendingTips` message the current admin can change the amount of pending tips that

Reviewer: ```suggestion
With the `UpdateMaxPendingTips` message, the current admin can change the amount of pending tips that
```

can be associated to a centralized application user.

### UpdateMaxSentPendingTips

> With the `UpdateMaxPendingTips` message the current admin can change the amount of tips that

Reviewer: ```suggestion
With the `UpdateMaxPendingTips` message, the current admin can change the amount of tips that
```

the contract collects for a specific sender address.

### RemovePendingTip

> With the `RemovePendingTip` message the user can cancel a tip that haven't been claimed yet.

Reviewer: ```suggestion
With the `RemovePendingTip` message, the user can cancel a tip that haven't been claimed yet.
```

    pub block_height: u64,
}
```

##### UnclaimedSentTips

> With the `UnclaimedSentTips` message the user can query the tips that has sent that aren't claimed yet.

Reviewer: ```suggestion
With the `UnclaimedSentTips` message, the user can query the unclaimed tips that has been sent.
```

The returned tips are provided as follows:
```rust
struct QueryUnclaimedSentTipsResponse {
    pub tips: Vec<Tip>,
}

struct Tip {
    pub sender: Addr,
    pub amount: Vec<Coin>,
    pub block_height: u64,
}
```

### Config

> With the  `Config{}` message a user can query the current contract's configurations.

Reviewer: ```suggestion
With the  `Config{}` message, a user can query the current contract's configurations.
```
